File: backtracking/backtracking.py
MAXCANDIDATES = 5

finished = False


def is_a_solution(a, k, n):
    return k == n


def construct_candidates(a, k, n, c, ncandidates):
    c[0] = 1
    c[1] = 0
    ncandidates[0] = 2


def process_solution(a, k):
    print("{ ", end='')
    for i in range(1, k+1):
        if a[i] == True:
            print(i, end=" ")
    print(" }")


def backtrack(a, k, input):
    c = [0]*MAXCANDIDATES
    ncandidates = [0]
    if(is_a_solution(a, k, input)):
        process_solution(a, k)
    else:
        k += 1
        construct_candidates(a, k, input, c, ncandidates)
        for i in range(ncandidates[0]):
            a[k] = c[i]
            backtrack(a, k, input)
            if(finished):
                return


a = [1, 2, 3]

backtrack(1, 0, 3)

File: backtracking/backtracking.cpp
#include <iostream>

using namespace std;

void backtrack(char[], int, int);
void swap(char[], int, int);
void process_solution(char[], int);

const int MAXCANDIDATES = 5;

int main()
{
    int k = 0, length = 3;
    char a[MAXCANDIDATES] = {'a', 'b', 'c'};
    backtrack(a, k, length);
}

void backtrack(char a[], int k, int n)
{
    if (k == n)
        process_solution(a, k);
    else
    {
        for (int i = k; i < n; i++)
        {
            swap(a, k, i);
            backtrack(a, k + 1, n);
            swap(a, k, i);
        }
    }
}
void swap(char a[], int i, int j)
{
    char temp;
    temp = a[i];
    a[i] = a[j];
    a[j] = temp;
}
void process_solution(char a[], int k)
{
    for (int i = 0; i < k; i++)
    {
        cout << " " << a[i];
    }
    cout << endl;
}
